src/jllvm/materialization/CodeGeneratorUtils.cpp

```diff
@@ -330,26 +364,46 @@ void ByteCodeTypeChecker::checkBasicBlock(llvm::ArrayRef<char> block, std::uint1
 
                 if (!type->isVoidTy())
                 {
-                    typeStack.push_back(type);
+                    typeStack.emplace_back(type);
+                }
+            },
+            [&](OneOf<JSR, JSRw> jsr)
+            {
+                std::uint16_t retAddress =
+                    jsr.offset + sizeof(OpCodes)
+                    + (holds_alternative<JSRw>(operation) ? sizeof(std::int32_t) : sizeof(std::int16_t));
+                std::uint16_t target = jsr.offset + jsr.target;
+
+                m_returnAddressToSubroutineMap.insert({retAddress, target});
+
+                if (auto iter = m_subroutineToReturnInfoMap.find(target); iter != m_subroutineToReturnInfoMap.end())
```

Comment:
Could you add a comment here what this `if`s purpose is?
Am I understanding correctly that you essentially "caching" and reusing the type stack at the end of the subroutine, if previously calculated, to avoid having to typecheck the subroutine more than once?